File: examples/predict.c
/* Binary anomaly labels from C. Build via `make cdemo-predict`. */
#include <stdio.h>
#include <stdlib.h>
#include "iforest.h"

int main(void)
{
    int n = 210, m = 2, i;
    double *X = malloc((size_t)n * m * sizeof(double));
    int *lab = malloc((size_t)n * sizeof(int));

    for (i = 0; i < 200; i++) {                 /* 200 inliers */
        X[i * m + 0] = (double)rand() / RAND_MAX;
        X[i * m + 1] = (double)rand() / RAND_MAX;
    }
    for (i = 200; i < n; i++) {                 /* 10 outliers */
        X[i * m + 0] = 20.0 + (double)rand() / RAND_MAX;
        X[i * m + 1] = 20.0 + (double)rand() / RAND_MAX;
    }

    iforest_t h = iforest_train(X, n, m, 100, 0);
    iforest_predict(h, X, n, m, lab, 0.6);

    int flagged = 0;
    for (i = 0; i < n; i++) flagged += lab[i];
    printf("flagged at 0.6 = %d (of %d; ~10 are outliers)\n", flagged, n);

    iforest_free(h);
    free(X);
    free(lab);
    return 0;
}

File: examples/predict.f90
! Binary anomaly labels, by fixed threshold and by contamination fraction.
program predict_example
  use iforest, only: dp, IsolationForest, train_forest, predict, free_forest
  implicit none
  type(IsolationForest) :: forest
  real(dp) :: X(210, 2)
  integer :: lab(210), i
  real(dp) :: r1, r2

  do i = 1, 200
     call random_number(r1); call random_number(r2)
     X(i,:) = [r1, r2]                        ! 200 inliers
  end do
  do i = 201, 210
     call random_number(r1); call random_number(r2)
     X(i,:) = [20.0_dp + r1, 20.0_dp + r2]    ! 10 outliers
  end do

  call train_forest(forest, X, 210)

  call predict(forest, X, 210, lab, threshold=0.6_dp)
  print '(a,i0)', "flagged at threshold 0.6:    ", sum(lab)

  call predict(forest, X, 210, lab, contamination=0.05_dp)
  print '(a,i0)', "flagged at 5% contamination: ", sum(lab)

  call free_forest(forest)
end program predict_example
